src/proto/streams/state.rs

```diff
@@ -217,10 +217,11 @@ impl State {
     pub fn recv_reset(&mut self, reason: Reason, queued: bool) {
 
         match self.inner {
-            Closed(Cause::EndStream) if queued => {
+            Closed(prior_reason) if queued => {
```

Comment:
this looks like the only functional change in this PR... can you explain how this resolves the issue?

Reply:
+1. It would be good to include a comment explaining this as it is definitely subtle.

Reply:
IIUC, the match can be condensed as:

```rust
    pub fn recv_reset(&mut self, reason: Reason, queued: bool) {
        match self.inner {
            Closed(..) if !queued => {},
            state => {
                trace!("recv_reset; {:?} => reason=Closed(Proto({:?})", state, reason);
                self.inner = Closed(Cause::Proto(reason));
            }
        }
}
```

I want to make sure I understand this properly, though.  What are the states where we can be closed and have queued data?

https://github.com/carllerche/h2/blob/b1d14caa2251cfa6316d73ad77409fb0be4a41f1/src/proto/streams/state.rs#L73-L87

In particular, I'm not sure if it makes sense to change the closed type when it's already `Proto` (it's probably harmless? but I'm not sure of the intended behavior).  Also, I'm not sure we'd want to override an `Io` error (though I'm not sure if it's actually possible to receive a reset after an Io error).

Reply:
This is fair, I can make the match more specific. 

My understanding is that we can have queued data in the following cases:
 - when the cause `Cause::Scheduled`,
- when the cause is `Cause::EndStream`, which would indicate that we've enqueued an EOS frame but it has yet to be sent (either because there are other frames ahead of it in the send queue, or we just haven't popped it yet)

AFAIK, `queued` should not be true with any other `Cause` variants, but I'll change the code to actually encode this nonetheless.

Reply:
@olix0r, hopefully 05a6017ba74d7561ad15ddc6d54363687a62db8f is clearer?